authorization.go

```diff
@@ -0,0 +1,204 @@
+	// the x509 client certificate.
+	IsAuthorized(clientOU []string, clientCN string) error
+
+	// IsAuthorizedForSite is called for requests which are accessing a resources
+	// belonging to a specific site.
+	// The resource is determined to belong to a site if the keyword `site` is present
+	// in it's `httprouter.Params`. In which case the value of this parameter is passed
+	// as `uriSite`.
+	// The values of `clientOU` and `clientCN` are the same as those of `IsAuthorized`.
+	IsAuthorizedForSite(clientOU []string, clientCN, uriSite string) error
+}
+
+// AllowOUs is a convenience function which produces an AuthorizationCheck from a list
+// of allowed OUs. See AllowSpecificOUandCNs for behavioral details.
+func AllowOUs(allowedOUs []string) AuthorizationCheck {
+	return AllowSpecificOUandCNs{OUs: allowedOUs, CNs: []string{}}
+}
+
+// AllowOUsWithSiteRestrictions is a convenience function which produces
+// two AuthorizationChecks which restrict all endpoints to a list of valid OUs,
+// and additional implement site-resource checks for a subset of those OUs.
+// AllowedOUs specifies which OUs a request must be part of to be serviced.
+// RestrictedSiteOUs specifies which OUs to apply site-resource checks to.
+// See AllowSpecificOUandCNs and RestrictSiteAccess for behavioral details on both
+// checks.
+// For example:
+//
+//   opts := certauth.Opts{
+//       AuthorizationChecks: AllowOUsWithSiteRestrictions(
+//           []string{"admin", "site", "employee"},
+//           []string{"site"},
+//       ),
+//   }
+//
+// This example would require *all* requests to present an OU of `admin`, `site`, or
+// `employee` and additionally, requests with an OU of `site` would be prevented from
+// accessing site-specific resources that were not part of their own site.
+func AllowOUsWithSiteRestrictions(
+	AllowedOUs, RestrictedSiteOUs []string,
+) []AuthorizationCheck {
+	return []AuthorizationCheck{
+		AllowOUs(AllowedOUs),
+		RestrictSiteAccess{AppliedOUs: RestrictedSiteOUs},
+	}
+}
+
+// AllowSpecificOUandCNs is an AuthorizationCheck which allows access to
+// resources for the specific Organizational Units and common names.
+// If any of the client's OUs match (i.e. `==`) any of the server's allowed OUs, *and*
+// the client's CN matches (i.e. `==`) one of the server's allowed CNs, the request
+// is allowed.
+// If `OUs` is empty, the client's OU is ignored, and only the CN is used to determine
+// authorization.
+// If `CNs` is empty, the client's CN is ignored, and only the OU is used to determine
+// authorization.
+// If both `OUs` and `CNs` are empty, no authorization check is performed.
+// Site resources are not considered specially. IsAuthorizedForSite has exactly the same
+// behavior as IsAuthorized (i.e. the site is ignored).
+type AllowSpecificOUandCNs struct {
+	OUs []string
+	CNs []string
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorized(
+	clientOU []string, clientCN string,
+) error {
```

Comment:
Why newline at all? I think 80 columns is too small for go, but 100 works great for me, and allows 3 vertical splits on one of my displays.

Reply:
> Is it gofmt/goimports compliant

Yes. All my code is:
```
autocmd BufWritePre *.go execute ':GoFmt'
```

90 columns is exactly half my window at 4k with a font size that's readable. That's why I wrapped several of these function definitions since they were just over my window width and the editor would wrap them anyway in a way that's much harder to read.

I'm not gonna argue that everyone should use 90 cause just cause that's what's convenient for me, but I'm gonna stick with that for my code so long as `gofmt` it. :)

Reply:
Good, just unused 😄